events/models.py

```diff
@@ -369,11 +357,38 @@ def send_free_spot_notifications_if_needed(self):
             obj.send_free_spot_notifications_if_needed()
 
     def upcoming(self):
-        return self.filter(time__gt=timezone.now())
+        return self.filter(time__gte=timezone.now())
+
+    def upcoming_with_leeway(self):
+        return self.filter(
+            time__gte=timezone.now()
+            - timedelta(minutes=settings.KUKKUU_ENROLLED_OCCURRENCE_IN_PAST_LEEWAY)
+        )
+
+    def upcoming_with_ongoing(self):
+        """Return occurrences that are upcoming or still ongoing (with added leeway)."""
+        qs = self.with_end_time()
+        return qs.filter(
+            end_time__gte=timezone.now()
+            - timedelta(minutes=settings.KUKKUU_ENROLLED_OCCURRENCE_IN_PAST_LEEWAY)
+        )
 
     def in_past(self):
         return self.exclude(time__gt=timezone.now())
 
+    def with_end_time(self):
```

Comment:
Good! Nice that Django ORM still casts this objects as Occurrence (and Enrolment).

Reply:
```
In [12]: e1 = Enrolment.objects.first()

In [13]: e2 = Enrolment.objects.with_end_time().first()

In [14]: e1.end_time
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-14-bfc454b61074> in <module>
----> 1 e1.end_time

AttributeError: 'Enrolment' object has no attribute 'end_time'

In [15]: e2.end_time
Out[15]: datetime.datetime(2001, 5, 31, 2, 49, 51, tzinfo=<UTC>)

In [16]: e1.__class__
Out[16]: events.models.Enrolment

In [17]: e2.__class__
Out[17]: events.models.Enrolment
```

Reply:
Sure it might be a bad thing in some case when a new model instance is created from a dict of another.